src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -109,7 +108,68 @@ def prepend(version: str) -> str:
     )
 
 
-def handle_dict_attr(proj_name: str, attributes: dict[str, str], fp: str) -> str:
+@dataclass(frozen=True)
+class PyProjectToml:
+    build_root: PurePath
+    toml_relpath: PurePath
+    toml_contents: str
+
+    @classmethod
+    def create(cls, parse_context: ParseContext, pyproject_toml_relpath: str) -> PyProjectToml:
+        build_root = Path(parse_context.build_root)
+        toml_relpath = PurePath(parse_context.rel_path, pyproject_toml_relpath)
+        return cls(
+            build_root=build_root,
+            toml_relpath=toml_relpath,
+            toml_contents=(build_root / toml_relpath).read_text(),
+        )
+
+    def parse(self) -> Mapping[str, Any]:
+        return toml.loads(self.toml_contents)
+
+    def _non_pants_project_abs_path(self, path: Path) -> Path | None:
+        resolved = path.resolve()
+        if resolved.is_file():
+            return resolved
+
+        try:
+            resolved.relative_to(self.build_root)
+        except ValueError:
+            return resolved
+
+        return None
+
+    def non_pants_project_abs_path(self, path: str) -> Path | None:
+        """Determine if the given path represents a non-Pants controlled project.
+
+        If the path points to a file, it's assumed the file is a distribution ( a wheel or sdist)
+        and the absolute path of that file is returned.
+
+        If the path points to a directory and that directory is outside of the build root, it's
+        assumed the directory is the root of a buildable Python project (i.e.: it contains a
+        pyproject.toml or setup.py) and the absolute path of the project is returned.
+
+        Otherwise, `None` is returned since the directory lies inside the build root and is assumed
+        to be a Pants controlled project.
+        """
```

Comment:
The resolution of filesystem paths makes this more susceptible to #7022, but I don't see a good way to resolve that without resolving the underlying issue.